### ISSUE TITLE
[MNG-7716] ConcurrencyDependencyGraph deadlock if no root can be selected

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#9265]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** opened **[MNG-7716](https://issues.apache.org/jira/browse/MNG-7716?redirect=false)** and commented

At Tycho we got a bug-report that results in a deadlock when calling the tycho-version-plugin:
https://github.com/eclipse-tycho/tycho/issues/2169

I debugged the problem and it seems that ConcurrencyDependencyGraph is actually the culprit here because it can return an empty list of projects to initially build. As no builds are scheduled then the code in the MultiThreadedBuilder waits forever for the result, this is what happening here:

* There is one segment ```
  org.faktorips:base:pom:23.6.0-SNAPSHOT -> [org.eclipse.tycho:tycho-versions-plugin:3.0.4-SNAPSHOT:set-version]

```
* This segment has a transitive reactor dependency to ```
MavenProject: org.faktorips:codequality-config:23.6.0-SNAPSHOT @ faktorips.base/codequality-config/pom.xml
``` (this is not a problem because we only execute a reactor=true mojo!)
* So now the loop finds that there is no "independent" project and returns an empty list
```

---

**Affects:** 3.8.7, 3.9.0, 4.0.0-alpha-4

**Remote Links:**
- [GitHub Pull Request #250
](https://github.com/apache/maven-integration-testing/pull/250)
- [GitHub Pull Request #1027
](https://github.com/apache/maven/pull/1027)
- [GitHub Pull Request #1028
](https://github.com/apache/maven/pull/1028)
- [GitHub Pull Request #1029
](https://github.com/apache/maven/pull/1029)

**Backported to:** [4.0.0-alpha-5](https://github.com/apache/maven/milestone/4?closed=1), [3.9.1](https://github.com/apache/maven/milestone/6?closed=1), [3.8.8](https://github.com/apache/maven/milestone/109?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Any change to have an IT here?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [f41d533e7133968f23664cf8cdde832dd305e16e](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=f41d533e7133968f23664cf8cdde832dd305e16e), with [837db7a756164cadd978e53f8c566f0f14b79b6c](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=837db7a756164cadd978e53f8c566f0f14b79b6c) for maven-3.9.x, with [a0f460ca5807bb98561e1ed123fd28302c4fa0f3](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=a0f460ca5807bb98561e1ed123fd28302c4fa0f3) for maven-3.8.x and ITs with [671cbfd274fa9516efb5a02b5cdd704e833f3d56](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git&a=commit&h=671cbfd274fa9516efb5a02b5cdd704e833f3d56).
